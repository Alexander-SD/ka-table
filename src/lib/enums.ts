export enum DataType {
  Boolean = 'boolean',
  Date = 'date',
  Number = 'number',
  Object = 'object',
  String = 'string',
}

export enum EditingMode {
  None = 'none',
  Cell = 'cell',
}

export enum ActionType {
  ChangeFilterRow = 'ChangeFilterRow',
  ChangeRowData = 'ChangeRowData',
  ChangeSorting = 'ChangeSorting',
  ChangeVirtualScrollingHeightSettings = 'ChangeVirtualScrollingHeightSettings',
  CloseEditor = 'CloseEditor',
  DeselectAllRows = 'DeselectAllRows',
  DeselectRow = 'DeselectRow',
  DeselectRowData = 'DeselectRowData',
  OpenEditor = 'OpenEditor',
  ScrollTable = 'ScrollTable',
  SelectAllRows = 'SelectAllRows',
  SelectSingleRow = 'SelectSingleRow',
  SelectRow = 'SelectRow',
  UpdateGroupsExpanded = 'UpdateGroupsExpanded',
}

export enum KeyboardEnum {
  Esc = 27,
  Enter = 13,
}

export enum SortDirection {
  Ascend = 'ascend',
  Descend = 'descend',
}

export enum SortingMode {
  None = 'none',
  Single = 'single',
}

export enum FilteringMode {
  None = 'none',
  FilterRow = 'filterRow',
}

export enum FilterOperatorName {
  Equal = '=',
  MoreThan = '>',
  LessThan = '<',
  MoreThanOrEqual = '>=',
  LessThanOrEqual = '<=',
  Contains = 'contains',
  IsEmpty = 'IsEmpty',
  IsNotEmpty = 'IsNotEmpty',
}
