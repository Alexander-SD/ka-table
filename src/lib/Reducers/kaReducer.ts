
import { ActionType } from '../enums';
import { ITableOption } from '../index';
import { Column } from '../models';
import { getCopyOfArrayAndInsertOrReplaceItem } from '../Utils/ArrayUtils';
import { addItemToEditableCells, removeItemFromEditableCells } from '../Utils/CellUtils';
import { replaceValue } from '../Utils/DataUtils';
import { getExpandedGroups, updateExpandedGroups } from '../Utils/GroupUtils';
import { getSortedColumns } from '../Utils/HeadRowUtils';
import { prepareTableOptions } from '../Utils/PropsUtils';

const kaReducer: any = (state: ITableOption, action: any) => {
  const {
    columns,
    data = [],
    editableCells = [],
    groupsExpanded,
    rowKeyField,
    selectedRows = [],
    virtualScrolling,
  } = state;

  switch (action.type) {
    case ActionType.OpenEditor: {
      const newEditableCells = addItemToEditableCells(
        action,
        editableCells);
      return { ...state, editableCells: newEditableCells };
    }
    case ActionType.CloseEditor: {
      const newEditableCells = removeItemFromEditableCells(
        action,
        editableCells);
      return { ...state, editableCells: newEditableCells };
    }
    case ActionType.UpdateFilterRowValue: {
      const column = columns.find((c: Column) => c.key === action.columnKey)!;
      const newColumn: Column = {
        ...column,
        filterRowValue: action.filterRowValue,
      };
      const newColumns = getCopyOfArrayAndInsertOrReplaceItem(
        newColumn,
        'key',
        columns,
      );
      return { ...state, columns: newColumns };
    }
    case ActionType.UpdateFilterRowOperator: {
      const column = columns.find((c: Column) => c.key === action.columnKey)!;
      const newColumn: Column = {
        ...column,
        filterRowOperator: action.filterRowOperator,
      };
      const newColumns = getCopyOfArrayAndInsertOrReplaceItem(
        newColumn,
        'key',
        columns,
      );
      return { ...state, columns: newColumns };
    }
    case ActionType.UpdateCellValue: {
      const row = data.find((d) => d[rowKeyField] === action.rowKeyValue);
      const column = columns.find((c) => c.key === action.columnKey)!;
      const newRowData = replaceValue(row, column, action.value);
      const newData = getCopyOfArrayAndInsertOrReplaceItem(newRowData, rowKeyField, data);
      return { ...state, data: newData };
    }
    case ActionType.DeleteRow: {
      const newData = data.filter(
        (d: any) => d[rowKeyField] !== action.rowKeyValue);
      return { ...state, data: newData };
    }
    case ActionType.SelectAllRows: {
      const newSelectedRows = data.map((d) => d[rowKeyField]);
      return { ...state, selectedRows: newSelectedRows };
    }
    case ActionType.Search: {
      return { ...state, search: action.search };
    }
    case ActionType.SelectSingleRow: {
      const newSelectedRows = [action.rowKeyValue];
      return { ...state, selectedRows: newSelectedRows };
    }
    case ActionType.DeselectAllRows:
      return { ...state, selectedRows: [] };
    case ActionType.SelectRow:
        return { ...state, selectedRows: [...selectedRows, ...[action.rowKeyValue]] };
    case ActionType.DeselectRow: {
      const newSelectedRows = [...selectedRows].filter((s) => s !== action.rowKeyValue);
      return { ...state, selectedRows: newSelectedRows };
    }
    case ActionType.UpdateSortingDirection:
      const sortedColumns = getSortedColumns(columns, action.columnKey);
      return { ...state, columns: sortedColumns };
    case ActionType.UpdateVirtualScrolling:
      return { ...state, virtualScrolling: action.virtualScrolling };
    case ActionType.UpdateData:
      return { ...state, data: action.data };
    case ActionType.ScrollTable:
      if (virtualScrolling) {
          const scrollTop = action.scrollTop;
          return {...state, ...{virtualScrolling: { ...virtualScrolling, scrollTop }}};
        }
      break;
    case ActionType.ExpandGroup: {
      let currentGroupsExpanded = groupsExpanded;
      if (!currentGroupsExpanded) {
        const preparedOptions = prepareTableOptions(state);
        currentGroupsExpanded = getExpandedGroups(preparedOptions.groupedData);
      }
      const newGroupsExpanded = updateExpandedGroups(
        currentGroupsExpanded,
        action.groupKey,
      );
      return { ...state, groupsExpanded: newGroupsExpanded };
    }
  }
  return state;
};

export {
  kaReducer,
};
