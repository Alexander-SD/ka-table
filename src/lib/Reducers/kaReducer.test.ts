import {
    deleteRow, deselectAllRows, deselectRow, selectAllRows, selectSingleRow, updateData
} from '../actionCreators';
import { ActionType } from '../enums';
import { kaReducer } from './kaReducer';

describe('kaReducer', () => {
  it('ScrollTable', () => {
    const scrollLeft = 10;
    const intialState = {
      columns: [],
      data: [],
      rowKeyField: '',
    };
    const newState = kaReducer(intialState, { type: ActionType.ScrollTable, scrollLeft });
    expect(newState).toEqual(intialState);
  });
  it('SelectAllRows', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
    };
    const newState = kaReducer(intialState, selectAllRows());
    expect(newState).toEqual({ ...intialState, selectedRows: [1, 2] });
  });
  it('SelectSingleRow', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
      selectedRows: [1],
    };
    const newState = kaReducer(intialState, selectSingleRow(2));
    expect(newState).toEqual({ ...intialState, selectedRows: [2] });
  });
  it('DeleteRow', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
    };
    const newState = kaReducer(intialState, deleteRow(2));
    expect(newState).toEqual({ ...intialState, data: [{ id: 1 }] });
  });
  it('DeselectAllRows', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
      selectedRows: [1, 2],
    };
    const newState = kaReducer(intialState, deselectAllRows());
    expect(newState).toEqual({ ...intialState, selectedRows: [] });
  });
  it('DeselectRow', () => {
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
      selectedRows: [1, 2],
    };
    const newState = kaReducer(intialState, deselectRow(2));
    expect(newState).toEqual({ ...intialState, selectedRows: [1] });
  });
  it('UpdateData', () => {
    const newData =  [{ id: 3 }, { id: 4 }];
    const intialState = {
      columns: [],
      data: [{ id: 1 }, { id: 2 }],
      rowKeyField: 'id',
      selectedRows: [1, 2],
    };
    const newState = kaReducer(intialState, updateData(newData));
    expect(newState.data).toEqual(newData);
  });
  it('ShowLoading', () => {
    const intialState = {
      loading: {
        enabled: false,
        text: 'test'
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.ShowLoading });
    expect(newState.loading.enabled).toEqual(true);
    expect(newState.loading.text).toEqual('test');
  });
  it('ShowLoading - change text', () => {
    const intialState = {
      loading: {
        enabled: false,
        text: 'test'
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.ShowLoading, text: '2' });
    expect(newState.loading.enabled).toEqual(true);
    expect(newState.loading.text).toEqual('2');
  });
  it('HideLoading', () => {
    const intialState = {
      loading: {
        enabled: true,
        text: 'test'
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.HideLoading });
    expect(newState.loading.enabled).toEqual(false);
    expect(newState.loading.text).toEqual('test');
  });
  it('HideLoading', () => {
    const intialState = {
      loading: {
        enabled: true,
        text: 'test'
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.HideLoading });
    expect(newState.loading.enabled).toEqual(false);
    expect(newState.loading.text).toEqual('test');
  });
  it('UpdatePageIndex', () => {
    const intialState = {
      paging: {
        enabled: true,
        pageIndex: 2,
      }
    };
    const newState = kaReducer(intialState, { type: ActionType.UpdatePageIndex, pageIndex: 5 });
    expect(newState.paging.enabled).toEqual(true);
    expect(newState.paging.pageIndex).toEqual(5);
  });
});
