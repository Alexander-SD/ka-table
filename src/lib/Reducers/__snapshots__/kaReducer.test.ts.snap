// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kaReducer ReorderRows default 1`] = `
Object {
  "data": Array [
    Object {
      "id": 2,
    },
    Object {
      "id": 3,
    },
    Object {
      "id": 1,
    },
    Object {
      "id": 4,
    },
  ],
  "rowKeyField": "id",
}
`;

exports[`kaReducer ReorderRows inner keys 1`] = `
Object {
  "data": Array [
    Object {
      "a": Object {
        "id": 2,
      },
    },
    Object {
      "a": Object {
        "id": 3,
      },
    },
    Object {
      "a": Object {
        "id": 1,
      },
    },
    Object {
      "a": Object {
        "id": 4,
      },
    },
  ],
  "rowKeyField": "a.id",
}
`;
